File: ios/Runner/Native/BatteryChannel.swift
import Flutter
import UIKit

final class BatteryChannel {

    private let channel: FlutterMethodChannel

    init(messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(
            name: "pdp.flutter.app/battery",
            binaryMessenger: messenger
        )

        channel.setMethodCallHandler { call, result in
            switch call.method {
            case "getBatteryLevel":
                if let batteryLevel = BatteryLevelReader.capacityPercent() {
                    result(batteryLevel)
                } else {
                    result(
                        FlutterError(
                            code: "UNAVAILABLE",
                            message: "Battery level not available.",
                            details: nil
                        )
                    )
                }

            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
}

File: android/app/src/main/kotlin/com/example/pdp_todo_app/native/BatteryChannel.kt
package com.example.pdp_todo_app.native

import android.content.Context
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

private const val BATTERY_CHANNEL = "pdp.flutter.app/battery"

class BatteryChannel(
    context: Context,
    messenger: BinaryMessenger
) {
    private val channel = MethodChannel(
        messenger,
        BATTERY_CHANNEL
    )

    init {
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "getBatteryLevel" -> {
                    val batteryLevel = BatteryLevelReader.capacityPercent(context)
                    if (batteryLevel != null) {
                        result.success(batteryLevel)
                    } else {
                        result.error(
                            "UNAVAILABLE",
                            "Battery level not available.",
                            null
                        )
                    }
                }

                else -> result.notImplemented()
            }
        }
    }
}
